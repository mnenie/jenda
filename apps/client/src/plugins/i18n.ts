import { createI18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '@/shared/libs/vee-validate/rules'
import { customPluralRule } from '@/shared/libs/i18n/plurals'
import { getCurrentLocale } from '@/shared/libs/i18n/utils/getLocale'

interface MessageSchema {
  [key: string]: any
}

const ru_ru = import.meta.glob(
  '@/shared/libs/i18n/locales/**/ru-RU.ts',
  { eager: true },
)
const en_us = import.meta.glob(
  '@/shared/libs/i18n/locales/**/en-US.ts',
  { eager: true },
)
const zh_cn = import.meta.glob(
  '@/shared/libs/i18n/locales/**/zh-CN.ts',
  { eager: true },
)

function loadLang(modules: Record<string, any>) {
  const messages: { [key: string]: string } = {}

  Object.keys(modules).forEach((module) => {
    Object.assign(messages, { ...modules[module].default })
  })
  return messages
}

const messages = {
  'en-US': {
    ...loadLang(en_us),
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...loadLang(ru_ru),
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...loadLang(zh_cn),
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

export default i18n
