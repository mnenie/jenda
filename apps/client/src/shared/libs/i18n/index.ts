import { createI18n, type I18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '../vee-validate/rules'
import { enLocale, ruLocale, zhLocale } from './locales'
import { customPluralRule } from './plurals'
import useDayJsLocalize from './dayjs'
import { getCurrentLocale } from './utils/getLocale'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

useDayJsLocalize(i18n as I18n)

export default i18n
