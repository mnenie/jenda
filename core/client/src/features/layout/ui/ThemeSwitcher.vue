<script setup lang="ts">
import { useDark, useToggle } from '@vueuse/core'
import { UiButton } from '@/shared/ui'

const isDark = useDark()
const toggleDark = useToggle(isDark)
</script>

<template>
  <UiButton
    variant="ghost"
    size="sm"
    class="p-2 gap-1 text-neutral-400 hover:text-neutral-700 dark:!text-neutral-400 dark:hover:!text-neutral-200"
    @click="toggleDark()"
  >
    <div v-if="!isDark" i-hugeicons-moon-02 text-base />
    <div v-else i-lucide-sun text-base />
  </UiButton>
</template>
