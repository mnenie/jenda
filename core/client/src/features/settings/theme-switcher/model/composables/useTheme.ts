import type { ShallowRef } from 'vue'
import type { ThemeBlock } from '..'
import { useColorMode } from '@vueuse/core'
import { ref } from 'vue'

export default function useTheme(themeBlocks: ShallowRef<ThemeBlock[]>) {
  const { store } = useColorMode()

  const selectedThemeId = ref(
    themeBlocks.value.find(item => item.active)?.id || 0,
  )

  const changeActiveTheme = (id: number) => {
    themeBlocks.value = themeBlocks.value.map((theme, index) => ({
      ...theme,
      active: index === id,
    }))
    selectedThemeId.value = id
  }

  const onNewPreferences = () => {
    const selectedTheme = themeBlocks.value.find(
      item => item.id === selectedThemeId.value,
    )
    if (selectedTheme) {
      store.value = selectedTheme.value
    }
  }

  return {
    changeActiveTheme,
    onNewPreferences,
  }
}
