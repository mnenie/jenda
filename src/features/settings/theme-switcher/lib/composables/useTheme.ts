import { ref } from 'vue';
import type { Ref } from 'vue';
import { useColorMode } from '@vueuse/core';
import type { ThemeBlock } from '../../model';

export default function useTheme(themeBlocks: Ref<ThemeBlock[]>) {
  const { store } = useColorMode();

  const selectedThemeId = ref(themeBlocks.value.find((item) => item.active)?.id || 0);

  const changeActiveTheme = (id: number) => {
    themeBlocks.value = themeBlocks.value.map((theme, index) => ({
      ...theme,
      active: index === id
    }));
    selectedThemeId.value = id;
  };

  const onNewPreferences = () => {
    const selectedTheme = themeBlocks.value.find((item) => item.id === selectedThemeId.value);
    if (selectedTheme) {
      store.value = selectedTheme.value;
    }
    console.log(store.value);
  };

  return {
    changeActiveTheme,
    onNewPreferences
  };
}
