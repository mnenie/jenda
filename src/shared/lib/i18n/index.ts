import { createI18n } from 'vue-i18n';
import { useCookies } from '@vueuse/integrations/useCookies';
import enLocale from './locales/en-US';
import ruLocale from './locales/ru-RU';
import zhLocale from './locales/zh-CN';

type MessageSchema = typeof enLocale;

const messages = {
  'en-US': {
    ...enLocale
  },
  'ru-RU': {
    ...ruLocale
  },
  'zh-CN': {
    ...zhLocale
  }
};
const cookies = useCookies();

const getCurrentLocale = () => {
  const cookieLanguage = cookies.get('i18n');
  if (cookieLanguage) {
    return cookieLanguage;
  }

  return 'en-US';
};

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale(),
  messages: messages
});

export default i18n;
