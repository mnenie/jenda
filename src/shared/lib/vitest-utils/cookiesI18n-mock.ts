import { vi } from 'vitest';

vi.mock('@vueuse/integrations/useCookies', () => {
  return {
    useCookies: () => ({
      get(key: string) {
        return key === 'i18n' ? 'en-US' : undefined;
      }
    })
  };
});
