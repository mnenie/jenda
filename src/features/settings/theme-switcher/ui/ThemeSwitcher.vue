<script setup lang="ts">
import { computed, shallowRef } from 'vue';
import { useColorMode } from '@vueuse/core';
import useTheme from '../model/composables/useTheme';
import type { ThemeBlock } from '../model';
import { UiButton, UiRadioGroupContainer } from '@/shared/ui';
import LightBlock from './LightBlock.vue';
import DarkBlock from './DarkBlock.vue';
import SystemBlock from './SystemBlock.vue';

const { store } = useColorMode();

const themeBlocks = shallowRef<ThemeBlock[]>([
  { id: 0, item: LightBlock, active: store.value === 'light', value: 'light' },
  { id: 1, item: DarkBlock, active: store.value === 'dark', value: 'dark' },
  { id: 2, item: SystemBlock, active: store.value === 'auto', value: 'auto' }
]);
const { onNewPreferences, changeActiveTheme } = useTheme(themeBlocks);

const themeAbout = computed(() => {
  return (value: string) => {
    if (value === 'auto') {
      return 'System';
    }
    return value
      .split('')
      .map((char, index) => (index === 0 ? char.toUpperCase() : char))
      .join('');
  };
});
</script>

<template>
  <form :class="$style.theme_form" @submit.prevent="onNewPreferences">
    <UiRadioGroupContainer>
      <div v-for="theme in themeBlocks" :key="theme.id" :class="$style.block">
        <component
          :is="theme.item"
          :class="{ [$style.active]: theme.active }"
          @click="changeActiveTheme(theme.id)"
        />
        <span class="text-xs">{{ themeAbout($t(`settings.theme.variants.${theme.value}`)) }}</span>
      </div>
    </UiRadioGroupContainer>
    <UiButton variant="outline" type="submit">{{ $t('settings.theme.btn') }}</UiButton>
  </form>
</template>

<style module lang="scss">
.theme_form {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 30px;

  .block {
    display: flex;
    flex-direction: column;
    align-items: center;
    gap: 5px;

    & > span {
      font-weight: 500 !important;
    }
  }
}
.active {
  border-color: var(--zinc-900);
  &:hover {
    border-color: var(--zinc-900);
  }
}

:global(html.dark) {
  .active {
    border-color: var(--zinc-300);
    &:hover {
      border-color: var(--zinc-300);
    }
  }
}
</style>
