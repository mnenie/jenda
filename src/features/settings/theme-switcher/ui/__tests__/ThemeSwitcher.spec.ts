import { describe, it, expect, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import { ref, shallowRef } from 'vue';
import '@/shared/lib/vitest-utils/cookiesI18n-mock';
import useTheme from '../../model/composables/useTheme';
import ThemeSwitcher from '../ThemeSwitcher.vue';
import i18n from '@/shared/lib/i18n';

vi.mock('@vueuse/core', () => ({
  useColorMode: vi.fn(() => ({
    store: ref('auto')
  }))
}));

describe('tests for ThemeSwitcher.vue', () => {
  const wrapper = mount(ThemeSwitcher, {
    global: {
      plugins: [i18n],
      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            'settings.theme.variants.light': 'light',
            'settings.theme.variants.dark': 'dark',
            'settings.theme.variants.auto': 'system'
          };
          return translations[key];
        }
      }
    }
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  it('should render subcomponents', () => {
    expect(wrapper.findComponent({ name: 'UiRadioGroupContainer' }).exists()).toBe(true);
    expect(wrapper.findComponent({ name: 'UiButton' }).exists()).toBe(true);
  });

  it('computed `themeAbout` is working correctly', () => {
    //@ts-expect-error PublicInstance
    const themeAbout = wrapper.vm.themeAbout;
    expect(themeAbout('light')).toBe('Light');
    expect(themeAbout('dark')).toBe('Dark');
    expect(themeAbout('auto')).toBe('System');
  });
});

describe('useTheme composable tests', () => {
  it('should change active theme correctly', () => {
    const themeBlocks = shallowRef([
      { id: 0, active: true, value: 'light' },
      { id: 1, active: false, value: 'dark' },
      { id: 2, active: false, value: 'auto' }
    ]);

    const { changeActiveTheme } = useTheme(themeBlocks as unknown as any);

    changeActiveTheme(1);
    expect(themeBlocks.value[0].active).toBe(false);
    expect(themeBlocks.value[1].active).toBe(true);
    expect(themeBlocks.value[2].active).toBe(false);
  });
});
