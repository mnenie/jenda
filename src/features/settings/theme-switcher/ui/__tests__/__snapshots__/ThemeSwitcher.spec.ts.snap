// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for ThemeSwitcher.vue > should render correctly 1`] = `
"<form class="theme_form">
  <ui-radio-group-container-stub></ui-radio-group-container-stub>
  <ui-button-stub variant="outline" size="md" loading="false" disabled="false" loadingplacement="leading" type="submit"></ui-button-stub>
</form>"
`;
